src/widgets/modals/ByProgramModal.js

```diff
@@ -1,376 +1,248 @@
+      const mosText = `${maxMOS}: ${item.maxMOS}`;
+      const thresholdText = `${threshMOS}: ${itemMOS}`;
+      const maxOrdersText = `${maxOPP}: ${maxOrders}`;
+      return `${mosText} - ${thresholdText} - ${maxOrdersText}`;
+    },
   },
-  misc: {
-    maxMOS: programStrings.max_mos,
-    thresholdMOS: programStrings.threshold_mos,
-    maxOrdersPerPeriod: programStrings.max_orders_per_period,
-    requisitions: programStrings.requisitions,
-    program: programStrings.program,
-    general: programStrings.general,
-    details: programStrings.details,
-    requisition: navStrings.requisition,
-    stocktake: navStrings.stocktake,
-  },
-  button: {
-    order: programStrings.order,
-    stocktake: generalStrings.stocktake,
+  period: {
+    onSelect: value => dispatch(selectPeriod(value)),
+    renderRightText: item => {
+      const { requisitions } = getLocalizedStrings();
+      const { maxOrdersPerPeriod } = orderType;
+      const requisitionsInPeriod = item.requisitionsForOrderType(program, orderType);
+      const periodText = `${requisitionsInPeriod}/${maxOrdersPerPeriod} ${requisitions}`;
+      return `${item.toString()} - ${periodText}`;
+    },
   },
 });
 
-const newState = {
-  RESET_ALL: {
-    program: null,
-    supplier: null,
-    period: null,
-    periods: null,
-    orderType: null,
-    orderTypes: null,
-    name: null,
-    currentStepKey: null,
-    isProgramBased: true,
-  },
-  SELECT_PROGRAM: {
-    period: {},
-    periods: null,
-    supplier: {},
-    orderType: {},
-    orderTypes: null,
-    modalIsOpen: false,
-  },
-  SELECT_ORDER_TYPE: { period: {}, periods: null, modalIsOpen: false },
-  SELECT_NAME: { modalIsOpen: false },
-  SELECT_PERIOD: { modalIsOpen: false },
-  SELECT_SUPPLIER: { modalIsOpen: false },
-};
-
-export class ByProgramModal extends React.Component {
-  constructor(props) {
-    super(props);
-
-    this.LOCALIZATION = getLocalizedStrings();
-    this.state = {
-      // Values used for deriving current UI state
-      steps: null,
-      modalIsOpen: false,
-      isProgramBased: true,
-      currentStepKey: null,
-      // New requisition or stocktake values
-      program: null,
-      supplier: null,
-      orderType: null,
-      period: null,
-      name: null,
-      // Data stored for performance
-      storeTag: null,
-      programs: null,
-      suppliers: null,
-      orderTypes: null,
-      periods: null,
-    };
-  }
-
-  // Store unchanging select modal options and set the default
-  // steps array.
-  componentDidMount = () => {
-    const { settings, database } = this.props;
-    const programs = getAllPrograms(settings, database);
-    const suppliers = database.objects('InternalSupplier');
-    this.setCurrentSteps(true);
-    this.setState({ programs, suppliers });
+export const ByProgramModal = ({ settings, database, type, onConfirm, ...props }) => {
+  const [state, dispatch] = useReducer(byProgramReducer, {}, () => initialState({ type }));
```

Comment:
Would be cool to make this a closure maybe?

```
getInitialiser({ type }) => () => ({
  program: null,
  supplier: null,
  ... 
  })
}
```

Then can do:

```
const initialiser = getInitialiser({ type })
const [state, dispatch] = useReducer(byProgramReducer, {}, initialiser);
```

Reply:
Prefer to keep it standard so that the reducer could potentially reset state by using the function normally

Reply:
👍 .